File: DriverAction.py
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.by import By
from typing import Union, List, Callable
from main import logger
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException as NSE
from functools import wraps
import random
import time

def wait_element_decorator(func: Callable) -> Callable:
    """
    Decorator to wait for a web element before executing the function.

    Parameters:
    -----------
    func : Callable
        The function to be decorated.

    Returns:
    --------
    Callable
        The wrapped function with element waiting logic.
    """
    @wraps(func)
    def wrapper(self, value: str, *args, wait_time: int = 20, _decorator_log: bool = False, by: By = None, **kwargs):
        by = self._by if by is None else by
        try:
            element = WebDriverWait(self._driver, wait_time).until(
                EC.presence_of_element_located((by, value))
            )
            if not element:
                raise NSE("Input element not found, please check By and make sure it is loaded correctly")
            if _decorator_log:
                logger.debug(f"Waited for element {value}")
        except Exception as e:
            logger.error(f"Error waiting for element {value}: {e}")
            raise
        return func(self, value, *args, log=kwargs.get('log', True), by=by, **kwargs)
    return wrapper

class DriverAction():
    """
    A class to perform various actions on web elements using Selenium WebDriver.

    Attributes:
    -----------
    _driver : WebDriver
        The Selenium WebDriver instance used to interact with the web browser.
    _by : By
        The method used to locate elements on the web page.

    Methods:
    --------
    click_element(value: str, elementname: str, log: bool = True) -> List[str]:
        Clicks on a web element and logs the action.

    double_click(value: str, elementname: str, log: bool = True, by: By = None) -> List[str]:
        Double-clicks on a web element and logs the action.

    right_click(value: str, elementname: str, log: bool = True, by: By = None) -> List[str]:
        Right-clicks on a web element and logs the action.

    get_element_attribute(value: str, attribute: str, log: bool = True, by: By = None) -> str:
        Retrieves the value of a specified attribute from a web element and logs the action.

    input_keys(value: str, log: bool = True, *keys: any, by: By = None) -> None:
        Sends keys to a web element and logs the action.

    wait_element(value: str, wait_time: int = 20, log: bool = True, by: By = None) -> None:
        Waits for a web element to be present on the web page and logs the action.
        
    slide_horizontal(self, value: str, offset: int, log: bool = True, by: By = None, slowly:bool = True) -> None:
        Slides a web element horizontally by a specified offset and logs the action.
    
    scroll_down(self, value:str = None, pixel:int = None, sleep_time:float = random.uniform(0.5, 1), log:bool = True, by: By = None)->None:
        Scrolls the web page down and logs the action.
    """

    def __init__(self, driver, by: By) -> None:
        self._driver = driver
        self._by = by
    

    @wait_element_decorator
    @logger.catch
    def click_element(self, value:str, elementname:str, log:bool = True, by: By = None) -> List[str]:
        by = self._by if by is None else by
        element = self._driver.find_element(by, value)
        element.click()
        if log:
            logger.debug(f"Clicked on element {elementname}")
        return self._driver.window_handles
    
    @wait_element_decorator
    @logger.catch
    def double_click(self, value: str, elementname: str, log:bool = True, by: By = None) -> List[str]:
        by = self._by if by is None else by
        element = self._driver.find_element(by, value)
        actions = ActionChains(self._driver)
        actions.double_click(element).perform()
        
        if log:
            logger.debug(f"Doubled clicked on element {elementname}")
        return self._driver.window_handles

    @wait_element_decorator
    @logger.catch
    def right_click(self, value: str, elementname: str, log:bool = True, by: By = None) -> List[str]:
        by = self._by if by is None else by
        element = self._driver.find_element(by, value)
        actions = ActionChains(self._driver)
        actions.context_click(element).perform()
        
        if log:
            logger.debug(f"Right clicked on element {elementname}")
        return self._driver.window_handles

    @wait_element_decorator
    @logger.catch
    def get_element_attribute(self, value:str, attribute:str, log:bool = True, by: By = None) -> str:
        by = self._by if by is None else by
        element = self._driver.find_element(by, value)
        result = element.get_attribute(attribute).strip()
        if log:
            logger.debug(f"Get attribute {attribute} on element, result: {result}")
        return result
    
    @wait_element_decorator
    @logger.catch
    def input_keys(self, value:str, *keys:any, log:bool = True, by: By = None) -> None:
        by = self._by if by is None else by
        element = self._driver.find_element(by, value)
        element.send_keys(*keys)
        if log:
            logger.debug(f"Input text {str(*keys)} into element{value}")

    @logger.catch
    def wait_element(self, value: str, wait_time: int = 20, log: bool = False, by: By = None) -> any:
        by = self._by if by is None else by
        element = WebDriverWait(self._driver, wait_time).until(
            EC.presence_of_element_located((by, value))
        )
        if not element:
            raise NSE("Input element not found, please check By and make sure it is loaded correctly")
        if log:
            logger.debug(f"Wait for element {value}")
        return element

    @wait_element_decorator
    @logger.catch
    def slide_horizontal(self, value: str, offset: int, log: bool = True, by: By = None, slowly: bool = True, slow_step:int = 10, slow_wait:float = 0.01) -> None:
        by = self._by if by is None else by
        element = self._driver.find_element(by, value)
        actions = ActionChains(self._driver)
        if not slowly:
            actions.click_and_hold(element).move_by_offset(offset, 0).release().perform()
        else:
            # move slowly
            actions.click_and_hold(element)
            i = 0
            step = slow_step 
            while i < offset:
                time.sleep(slow_wait)
                move_step = step if (offset - i) >= step else (offset - i)
                actions.move_by_offset(move_step, 0).perform()
                i += move_step
            actions.release().perform() 

        if log:
            logger.debug(f"Slide element {value} by offset {offset}")


    @logger.catch
    def scroll_down(self, value:str = None, pixel:int = None, sleep_time:float = random.uniform(0.5, 1), log:bool = True, by: By = None, slowly: bool = True, slow_step:int = 100) -> None:
        """
        value: a By expression for element search, then driver will scroll until it is in view
        pixel: how many pixel to scroll down

        if none of the first two are provided, the page will be scrolled to the bottom gradually
        """
        by = self._by if by is None else by
        if pixel:
            """
            this option is rarely used and required to wait until the loading finishes
            """
            time.sleep(sleep_time * 4) # not optimal, change on condition
            if slowly:
                    for i in range(0, pixel, slow_step):
                        scroll_amount = slow_step if (pixel - i) >= slow_step else pixel - i
                        self._driver.execute_script('window.scrollBy(0, {})'.format(str(scroll_amount)))
                        time.sleep(sleep_time)
            else:
                self._driver.execute_script('window.scrollBy(0, {})'.format(str(pixel)))
            if log:
                logger.debug(f"Scroll down {pixel} pixel")

        elif value:
            self.wait_element(value, by=by)
            element = self._driver.find_element(by, value)

            if slowly:
                current_position = self._driver.execute_script("return window.pageYOffset;")
                target_position = element.location['y']
                # 200 works as a buffer
                target = target_position - 200 if target_position > 200 else 0  

                while current_position < target:
                    current_position += slow_step
                    if current_position > target:
                        current_position = target
                    self._driver.execute_script(f"window.scrollTo(0, {current_position});")
                    time.sleep(sleep_time)
            else:
                self._driver.execute_script("arguments[0].scrollIntoView();", element)

            if log:
                logger.debug(f"Scroll down to element {value}")
        else:
            if slowly:
                js = "return action=document.body.scrollHeight"
                height = 0
                new_height = self._driver.execute_script(js)
                while height < new_height:
                    for i in range(height, new_height, slow_step):
                        self._driver.execute_script('window.scrollTo(0, {})'.format(i))
                        time.sleep(sleep_time)
                    height = new_height
                    new_height = self._driver.execute_script(js)
            else:
                """
                This action is easily detected by the website and must wait until the scroll bar is loaded, 
                thus not recommended.
                """
                time.sleep(sleep_time * 4) # not optimal, change on condition
                self._driver.execute_script("var q=document.documentElement.scrollTop=10000")

            if log:

                logger.debug(f"Scroll down to the bottom")

    """
    awaiting test
    """
    @logger.catch
    def add_cookies(self, cookieinstance: Union[dict, List[dict]], log: bool = True) -> None:
        if isinstance(cookieinstance, dict):
            self._driver.add_cookie(cookieinstance)
            if 'expiry' in cookieinstance:
                del cookieinstance['expiry']
            if log:
                logger.debug(f"Added cookie: {cookieinstance}")
        elif isinstance(cookieinstance, list):
            for cookie in cookieinstance:
                self._driver.add_cookie(cookie)
                if 'expiry' in cookie:
                    del cookie['expiry']
                if log:
                    logger.debug(f"Added cookie: {cookie}")
    
    """
    awaiting test
    """
    @logger.catch
    def window_switch(self, ActionList: List[Union[int, tuple]], log: bool = True, by: By = None) -> None:
        """
        This function works as a second layer for abstract workflow,
        packing up a chain of action in ActionList for execution.
        An action can be a window index or a function for taking element etc.
        """
        by = self._by if by is None else by
        for action in ActionList:
            if isinstance(action, int):
                self._driver.switch_to.window(action)
                if log:
                    logger.debug(f"Switched to window {self._driver.title}")
            elif isinstance(action, tuple):
                func, args, kwargs = action
                func(*args, **kwargs)
            else:
                raise ValueError("ActionList items must be either int or tuple(func, args, kwargs)")
        if log:
            logger.debug("Window switch completed")

    @logger.catch
    def frame_switch(self, ActionList: List[Union[str, tuple]], log: bool = True, by: By = None) -> None:
        """
        This function works as a second layer for abstract workflow,
        packing up a chain of actions in ActionList for execution.
        An action can be a frame locator or a tuple containing a function and its arguments.
        """
        by = self._by if by is None else by
        for action in ActionList:
            if isinstance(action, str):
                self.wait_element(action, by=by)
                element = self._driver.find_element(by, action)
                self._driver.switch_to.frame(element)
                if log:
                    logger.debug(f"Switched to frame {action}")
            elif isinstance(action, tuple):
                func, args= action
                func(*args)
            else:
                raise ValueError("ActionList items must be either str or tuple(func, args, kwargs)")
        if log:
            logger.debug("Frame switch completed")

    @staticmethod
    @logger.catch
    def driver_signiture_validate(driver):
        """
        Validates the Selenium WebDriver's signature by navigating to bot.sannysoft.com,
        which is specially made for signiture test.

        Parameters:
        -----------
        driver : WebDriver
            The Selenium WebDriver instance used to interact with the web browser.

        Returns:
        --------
        None
            Logs the result of the validation process. If all checks pass, logs a success message.
            Otherwise, logs warnings for each failed check.
        """
        base_url = "https://bot.sannysoft.com/"
        driver.get(base_url)
        all_pass = True
        WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.XPATH, '//*[@id="fp2"]/tr[20]/td[2]')))
        tds = driver.find_elements(by=By.XPATH, value='//*[@id="fp2"]/tr/td[2]')
        for td in tds:
            if td.text != "ok":
                all_pass = False
                tr = driver.find_element(by=By.XPATH, value='//*[@id="fp2"]/tr[{}]/td[1]'.format(tds.index(td) + 1))
                logger.warning(f"Selenium driver signiture test failed in: {tr.text}, type: {td.text}")
        if all_pass:
            logger.success("Selenium driver signiture passed")

File: src/main/java/selenium/DriverAction.java
package selenium;


import org.openqa.selenium.*;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.List;
import java.util.Map;

public class DriverAction {
    private WebDriver driver;
    private static final Logger logger = LoggerFactory.getLogger(DriverAction.class);

    public DriverAction(WebDriver driver, By by, Map<String, Object> contact, String emailLevel) {
        this.driver = driver;
    }

    /**
     * Waits for a web element to be present.
     *
     * @param locator       The By locator of the element.
     * @param waitTime      Maximum time to wait in seconds.
     * @param decoratorLog  Whether to log the waiting action.
     * @return The found WebElement.
     */
    private WebElement wait(By locator, int waitTime, boolean decoratorLog) {
        try {
            WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(waitTime));
            WebElement element = wait.until(ExpectedConditions.presenceOfElementLocated(locator));
            if (decoratorLog) {
                logger.debug("Waited for element {}", locator);
            }
            return element;
        } catch (Exception e) {
            logger.error("Error waiting for element {}: {}", locator, e.getMessage());
            throw e;
        }
    }

    /**
     * Clicks on a web element.
     *
     * @param locator      The By locator of the element.
     * @param elementName  The name of the element for logging.
     * @param log          Whether to log the action.
     * @return A list of window handles.
     */
    public List<String> clickElement(By locator, String elementName, boolean log) {
        try {
            WebElement element = wait(locator, 20, true);
            element.click();
            if (log) {
                logger.debug("Clicked on element {}", elementName);
            }
            return List.copyOf(driver.getWindowHandles());
        } catch (Exception e) {
            logger.error("Failed to click element {}: {}", elementName, e.getMessage());
            throw e;
        }
    }

    /**
     * Double-clicks on a web element.
     *
     * @param locator      The By locator of the element.
     * @param elementName  The name of the element for logging.
     * @param log          Whether to log the action.
     * @return A list of window handles.
     */
    public List<String> doubleClick(By locator, String elementName, boolean log) {
        try {
            WebElement element = wait(locator, 20, true);
            Actions actions = new Actions(driver);
            actions.doubleClick(element).perform();
            if (log) {
                logger.debug("Double clicked on element {}", elementName);
            }
            return List.copyOf(driver.getWindowHandles());
        } catch (Exception e) {
            logger.error("Failed to double click element {}: {}", elementName, e.getMessage());
            throw e;
        }
    }

    /**
     * Right-clicks on a web element.
     *
     * @param locator      The By locator of the element.
     * @param elementName  The name of the element for logging.
     * @param log          Whether to log the action.
     * @return A list of window handles.
     */
    public List<String> rightClick(By locator, String elementName, boolean log) {
        try {
            WebElement element = wait(locator, 20, true);
            Actions actions = new Actions(driver);
            actions.contextClick(element).perform();
            if (log) {
                logger.debug("Right clicked on element {}", elementName);
            }
            return List.copyOf(driver.getWindowHandles());
        } catch (Exception e) {
            logger.error("Failed to right click element {}: {}", elementName, e.getMessage());
            throw e;
        }
    }

    /**
     * Retrieves the value of a specified attribute from a web element.
     *
     * @param locator    The By locator of the element.
     * @param attribute  The attribute name.
     * @param log        Whether to log the action.
     * @return The attribute value.
     */
    public String getElementAttribute(By locator, String attribute, boolean log) {
        try {
            WebElement element = wait(locator, 20, true);
            String result = element.getAttribute(attribute).trim();
            if (log) {
                logger.debug("Get attribute " + attribute + " on element " + locator + ", result: " + result);
            }
            return result;
        } catch (Exception e) {
            logger.error("Failed to get attribute" + attribute + "from element "+locator+": " + e.getMessage());
            throw e;
        }
    }

    /**
     * Sends keys to a web element.
     *
     * @param locator  The By locator of the element.
     * @param keys     The keys to send.
     * @param log      Whether to log the action.
     */
    public void inputKeys(By locator, String keys, boolean log) {
        try {
            WebElement element = wait(locator, 20, true);
            element.sendKeys(keys);
            if (log) {
                logger.debug("Input text '{}' into element {}", keys, locator);
            }
        } catch (Exception e) {
            logger.error("Failed to input keys into element {}: {}", locator, e.getMessage());
            throw e;
        }
    }

    /**
     * Waits for a web element to be present.
     *
     * @param locator  The By locator of the element.
     * @param waitTime Maximum time to wait in seconds.
     * @param log      Whether to log the action.
     */
    public void waitElement(By locator, int waitTime, boolean log) {
        try {
            waitElement(locator, waitTime, log);
            if (log) {
                logger.debug("Waited for element {}", locator);
            }
        } catch (Exception e) {
            logger.error("Failed to wait for element {}: {}", locator, e.getMessage());
            throw e;
        }
    }

    /**
     * Slides a web element horizontally by a specified offset.
     *
     * @param locator      The By locator of the element.
     * @param offset       The horizontal offset.
     * @param log          Whether to log the action.
     * @param slowly       Whether to perform the action slowly.
     * @param slowStep     The step size for slow movement.
     * @param slowWait     The wait time between steps in seconds.
     */
    public void slideHorizontal(By locator, int offset, boolean log, boolean slowly, int slowStep, double slowWait) {
        try {
            WebElement element = wait(locator, 20, true);
            Actions actions = new Actions(driver);
            if (!slowly) {
                actions.clickAndHold(element).moveByOffset(offset, 0).release().perform();
            } else {
                actions.clickAndHold(element).perform();
                int i = 0;
                while (i < offset) {
                    try {
                        Thread.sleep((long) (slowWait * 1000));
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        logger.error("Thread interrupted: {}", e.getMessage());
                        throw new RuntimeException(e);
                    }
                    int moveStep = Math.min(slowStep, offset - i);
                    actions.moveByOffset(moveStep, 0).perform();
                    i += moveStep;
                }
                actions.release().perform();
            }
            if (log) {
                logger.debug("Slide element {} by offset {}", locator, offset);
            }
        } catch (Exception e) {
            logger.error("Failed to slide horizontally on element {}: {}", locator, e.getMessage());
            throw e;
        }
    }

    /**
     * Scrolls the web page down.
     *
     * @param locator    The By locator of the target element (optional).
     * @param pixel      The number of pixels to scroll (optional).
     * @param sleepTime  The sleep time in seconds between scrolls.
     * @param log        Whether to log the action.
     * @param slowly     Whether to perform the scroll slowly.
     * @param slowStep   The step size for slow scrolling.
     */
    public void scrollDown(By locator, Integer pixel, double sleepTime, boolean log, boolean slowly, int slowStep) {
        try {
            JavascriptExecutor js = (JavascriptExecutor) driver;
            if (pixel != null) {
                Thread.sleep((long) (sleepTime * 4000));
                if (slowly) {
                    for (int i = 0; i < pixel; i += slowStep) {
                        int scrollAmount = Math.min(slowStep, pixel - i);
                        js.executeScript("window.scrollBy(0, " + scrollAmount + ");");
                        Thread.sleep((long) (sleepTime * 1000));
                    }
                } else {
                    js.executeScript("window.scrollBy(0, " + pixel + ");");
                }
                if (log) {
                    logger.debug("Scroll down {} pixels", pixel);
                }
            } else if (locator != null) {
                WebElement element = wait(locator, 20, true);
                if (slowly) {
                    Long currentPosition = (Long) js.executeScript("return window.pageYOffset;");
                    Long targetPosition = (long) element.getLocation().getY();
                    targetPosition = targetPosition > 200 ? targetPosition - 200 : 0;
                    while (currentPosition < targetPosition) {
                        currentPosition += slowStep;
                        if (currentPosition > targetPosition) {
                            currentPosition = targetPosition;
                        }
                        js.executeScript("window.scrollTo(0, " + currentPosition + ");");
                        Thread.sleep((long) (sleepTime * 1000));
                    }
                } else {
                    js.executeScript("arguments[0].scrollIntoView();", element);
                }
                if (log) {
                    logger.debug("Scroll down to element {}", locator);
                }
            } else {
                if (slowly) {
                    Long height = (Long) js.executeScript("return document.body.scrollHeight");
                    Long newHeight = height;
                    while (height < newHeight) {
                        for (long i = height; i < newHeight; i += slowStep) {
                            js.executeScript("window.scrollTo(0, " + i + ");");
                            Thread.sleep((long) (sleepTime * 1000));
                        }
                        height = newHeight;
                        newHeight = (Long) js.executeScript("return document.body.scrollHeight");
                    }
                } else {
                    Thread.sleep((long) (sleepTime * 4000));
                    js.executeScript("window.scrollTo(0, 10000);");
                }
                if (log) {
                    logger.debug("Scroll down to the bottom");
                }
            }
        } catch (Exception e) {
            logger.error("Failed to scroll down: {}", e.getMessage());
            throw new RuntimeException(e);
        }
    }

    /**
     * Adds cookies to the browser.
     *
     * @param cookieInstance The cookie(s) to add. Can be a single Map or a List of Maps.
     * @param log            Whether to log the action.
     */
    @SuppressWarnings("unchecked")
    public void addCookies(Object cookieInstance, boolean log) {
        try {
            if (cookieInstance instanceof Map) {
                Map<String, Object> cookie = (Map<String, Object>) cookieInstance;
                addSingleCookie(cookie, log);
            } else if (cookieInstance instanceof List) {
                List<Map<String, Object>> cookies = (List<Map<String, Object>>) cookieInstance;
                for (Map<String, Object> cookie : cookies) {
                    addSingleCookie(cookie, log);
                }
            }
        } catch (Exception e) {
            logger.error("Failed to add cookies: {}", e.getMessage());
            throw e;
        }
    }

    /**
     * Helper method to add a single cookie.
     *
     * @param cookie The cookie data.
     * @param log    Whether to log the action.
     */
    private void addSingleCookie(Map<String, Object> cookie, boolean log) {
        try {
            Cookie.Builder builder = new Cookie.Builder(cookie.get("name").toString(), cookie.get("value").toString());
            if (cookie.containsKey("domain")) {
                builder.domain(cookie.get("domain").toString());
            }
            if (cookie.containsKey("path")) {
                builder.path(cookie.get("path").toString());
            }
            if (cookie.containsKey("expiry")) {
                builder.expiresOn(new java.util.Date((Long) cookie.get("expiry") * 1000));
            }
            if (Boolean.TRUE.equals(cookie.get("secure"))) {
                builder.isSecure(true);
            }
            driver.manage().addCookie(builder.build());
            cookie.remove("expiry"); // Remove 'expiry' as per original Python logic
            if (log) {
                logger.debug("Added cookie: {}", cookie);
            }
        } catch (Exception e) {
            logger.error("Failed to add single cookie {}: {}", cookie, e.getMessage());
            throw e;
        }
    }

    /**
     * Switches between browser windows based on the provided action list.
     *
     * @param actionList The list of actions to perform. Each action can be an Integer (window index) or a Runnable.
     * @param log        Whether to log the actions.
     * @return The title of the current window after switching.
     */
    public String windowSwitch(List<Object> actionList, boolean log) {
        try {
            List<String> windowHandles = List.copyOf(driver.getWindowHandles());
            for (Object action : actionList) {
                if (action instanceof Integer) {
                    int index = (Integer) action;
                    if (index >= 0 && index < windowHandles.size()) {
                        driver.switchTo().window(windowHandles.get(index));
                        if (log) {
                            logger.debug("Switched to window {}", driver.getTitle());
                        }
                    } else {
                        throw new IndexOutOfBoundsException("Window index out of range: " + index);
                    }
                } else if (action instanceof Runnable) {
                    Runnable func = (Runnable) action;
                    func.run();
                } else {
                    throw new IllegalArgumentException("ActionList items must be either Integer or Runnable");
                }
            }
            if (log) {
                logger.debug("Window switch completed");
            }
            return driver.getTitle();
        } catch (Exception e) {
            logger.error("Failed to switch windows: {}", e.getMessage());
            throw e;
        }
    }

    /**
     * Switches between frames based on the provided action list.
     *
     * @param actionList The list of actions to perform. Each action can be a String (frame locator) or a Runnable.
     * @param log        Whether to log the actions.
     * @param customBy   The By locator to use for frame elements.
     */
    public void frameSwitch(List<Object> actionList, boolean log, By customBy) {
        try {
            for (Object action : actionList) {
                if (action instanceof String) {
                    By locator = ByLocator(customBy, (String) action);
                    WebElement frameElement = wait(locator, 20, true);
                    driver.switchTo().frame(frameElement);
                    if (log) {
                        logger.debug("Switched to frame {}", action);
                    }
                } else if (action instanceof Runnable) {
                    Runnable func = (Runnable) action;
                    func.run();
                } else {
                    throw new IllegalArgumentException("ActionList items must be either String or Runnable");
                }
            }
            if (log) {
                logger.debug("Frame switch completed");
            }
        } catch (Exception e) {
            logger.error("Failed to switch frames: {}", e.getMessage());
            throw e;
        }
    }

    /**
     * Helper method to reconstruct a By locator from a base locator and a value.
     *
     * @param baseBy The base By locator (e.g., By.id, By.name).
     * @param value  The value to locate the element.
     * @return A new By locator with the updated value.
     */
    private By ByLocator(By baseBy, String value) {
        String byType = baseBy.toString().split(":")[0].trim();
        switch (byType) {
            case "By.id":
                return By.id(value);
            case "By.name":
                return By.name(value);
            case "By.xpath":
                return By.xpath(value);
            case "By.cssSelector":
                return By.cssSelector(value);
            case "By.className":
                return By.className(value);
            case "By.linkText":
                return By.linkText(value);
            case "By.partialLinkText":
                return By.partialLinkText(value);
            case "By.tagName":
                return By.tagName(value);
            default:
                throw new IllegalArgumentException("Unsupported By type: " + byType);
        }
    }

    /**
     * Validates the Selenium WebDriver's signature by navigating to bot.sannysoft.com.
     *
     * @param driver The Selenium WebDriver instance.
     */
    public static void driverSignatureValidate(WebDriver driver) {
        Logger logger = LoggerFactory.getLogger(DriverAction.class);
        try {
            String baseUrl = "https://bot.sannysoft.com/";
            driver.get(baseUrl);
            WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
            wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath("//*[@id=\"fp2\"]/tr[20]/td[2]")));
            List<WebElement> tds = driver.findElements(By.xpath("//*[@id=\"fp2\"]/tr/td[2]"));
            boolean allPass = true;
            for (int i = 0; i < tds.size(); i++) {
                WebElement td = tds.get(i);
                if (!"ok".equalsIgnoreCase(td.getText())) {
                    allPass = false;
                    WebElement tr = driver.findElement(By.xpath("//*[@id=\"fp2\"]/tr[" + (i + 1) + "]/td[1]"));
                    logger.warn("Selenium driver signature test failed in: {}, type: {}", tr.getText(), td.getText());
                }
            }
            if (allPass) {
                logger.info("Selenium driver signature passed");
            }
        } catch (Exception e) {
            logger.error("Driver signature validation failed: {}", e.getMessage());
            throw e;
        }
    }
}